babel/src/main/codegen/config.fmpp

```diff
@@ -833,6 +835,7 @@ data: {
       # Return type of method implementation should be "SqlIdentifier".
       # Example: SqlParseTimeStampZ().
       dataTypeParserMethods: [
+         "PostgreSQLTypes"
```

Comment:
I'm on the fence about calling this `PostgreSQLTypes`. Any thoughts @julianhyde?

Reply:
You're concern is valid since the `TEXT` type exists for other DB types (ex: MySQL). I'll wait for suggestions and tell you if I come up with something else.

Reply:
A lot of databases support user-defined types. In Postgres I can say `(1 + 2) :: Foo` if `Foo` is a user-defined type.

Why not treat `TEXT` as if it is a user-defined type? The parser will just see an identifier.

Reply:
Because the metadata for the returned resultset for that query using calcite won't return `Foo` for the column type name, it will return whatever `SqlTypeName` `Foo` was configured to act like (ex: `BIGING`).

**Example:**
The following code prints `TEXT`
```java
Connection connection = // A connection to a real PostgreSQL DB
PreparedStatement statement = connection.prepareStatement("SELECT column1::TEXT FROM (VALUES (1, 2, 3)) a");
System.out.println(statement.executeQuery().getMetaData().getColumnTypeName(1));
```

The following code prints `VARCHAR`
```java
Connection connection = DriverManager.getConnection("jdbc:calcite:", info);
connection.unwrap(CalciteConnection.class).getRootSchema().unwrap(CalciteSchema.class).add("TEXT", new RelProtoDataType() {
            @Override
            public RelDataType apply(RelDataTypeFactory factory) {
                 return factory.createTypeWithNullability(factory.createSqlType(SqlTypeName.VARCHAR), false);
            }
        });

PreparedStatement statement = connection.prepareStatement("SELECT CAST(EXPR$0 AS text) FROM (VALUES (1, 2, 3)) a");
System.out.println(statement.executeQuery().getMetaData().getColumnTypeName(1));
```

Reply:
Because the metadata for the returned resultset for that query using calcite won't return `Foo` for the column type name, it will return whatever `SqlTypeName` `Foo` was configured to act like (ex: `BIGING`).

**Example:**
The following code prints `TEXT`
```java
Connection connection = // A connection to a real PostgreSQL DB
PreparedStatement statement = connection.prepareStatement("SELECT column1::TEXT FROM (VALUES (1, 2, 3)) a");
System.out.println(statement.executeQuery().getMetaData().getColumnTypeName(1));
```

The following code prints `VARCHAR`
```java
Connection connection = DriverManager.getConnection("jdbc:calcite:", info);
connection.unwrap(CalciteConnection.class).getRootSchema().unwrap(CalciteSchema.class).add("TEXT", new RelProtoDataType() {
            @Override
            public RelDataType apply(RelDataTypeFactory factory) {
                 return factory.createTypeWithNullability(factory.createSqlType(SqlTypeName.VARCHAR), false);
            }
        });

PreparedStatement statement = connection.prepareStatement("SELECT CAST(EXPR$0 AS text) FROM (VALUES (1, 2, 3)) a");
System.out.println(statement.executeQuery().getMetaData().getColumnTypeName(1));
```

Reply:
It's not surprising that the JDBC driver is returning VARCHAR, because your code is converting TEXT to the built-in VARCHAR type. How about converting TEXT to a user-defined type?

This really is not a parser issue.

Reply:
I'm convinced this is not a parser issue after I read your comment on the Jira issue. But I don't understand what you mean by 
> How about converting TEXT to a user-defined type?
I mean how can I do that ? Is there an example ?

Reply:
The names seems too broad, if you just want to define a `TEXT` type, then `PostSqlTextType` seems better name.

Reply:
I thought having a single method for all PostgreSQL types would be more reusable. But @michaelmior raised a valid flag since `TEXT` is not used by PostgreSQL only.
Do you happen to know how different is what I did and what @julianhyde is asking ?

Reply:
To the parser, a user-defined type is simply an identifier. No parser changes needed.

Reply:
So how exactly can I acheive what I mentioned [here](https://github.com/apache/calcite/pull/1250#discussion_r290965310) without modifying the parser ?

Reply:
You said that I need to define a user-defined type and that's what I did. So is there is even a way ?